feature_extractor.py

```diff
@@ -0,0 +1,292 @@
+            *interleave(conv_extract, [nn.ELU()] * len(conv_extract))
+        )
+
+        # CNN crop model.
+        conv_extract_crop = [
+            nn.Conv2d(
+                in_channels=in_channels[i],
+                out_channels=out_channels[i],
+                kernel_size=(F, F),
+                stride=S,
+                padding=P,
+            )
+            for i in range(L)
+        ]
+
+        self.extract_crop_representation = nn.Sequential(
+            *interleave(conv_extract_crop, [nn.ELU()] * len(conv_extract))
+        )
+
+        out_dim = self.k_dim
+        # CNN over full glyph map
+        out_dim += self.H * self.W * Y
+
+        # CNN crop model.
+        out_dim += self.crop_dim**2 * Y
+
+        self.embed_blstats = nn.Sequential(
+            nn.Linear(self.blstats_size, self.k_dim),
+            nn.ReLU(),
+            nn.Linear(self.k_dim, self.k_dim),
+            nn.ReLU(),
+        )
+
+        self.fc = nn.Sequential(
+            nn.Linear(out_dim, self.h_dim),
+            nn.ReLU(),
+            nn.Linear(self.h_dim, self.h_dim),
+            nn.ReLU(),
+        )
+
+        if self.use_lstm:
+            self.core = nn.LSTM(self.h_dim, self.h_dim, num_layers=1)
+
+        self.policy = nn.Linear(self.h_dim, self.num_actions)
+        self.baseline = nn.Linear(self.h_dim, 1)
+
+    def initial_state(self, batch_size=1):
+        if not self.use_lstm:
+            return tuple()
+        return tuple(
+            torch.zeros(self.core.num_layers, batch_size, self.core.hidden_size)
+            for _ in range(2)
+        )
+
+    def _select(self, embed, x):
+        # Work around slow backward pass of nn.Embedding, see
+        # https://github.com/pytorch/pytorch/issues/24912
+        out = embed.weight.index_select(0, x.reshape(-1))
+        return out.reshape(x.shape + (-1,))
+
+    def forward(self, env_outputs, core_state, done):
+        # -- [T x B x H x W]
+        glyphs = env_outputs["glyphs"]
+
+        # -- [T x B x F]
+        blstats = env_outputs["blstats"]
+
+        T, B, *_ = glyphs.shape
+
+        # -- [B' x H x W]
+        glyphs = torch.flatten(glyphs, 0, 1)  # Merge time and batch.
+
+        # -- [B' x F]
+        blstats = blstats.view(T * B, -1).float()
+
+        # -- [B x H x W]
+        glyphs = glyphs.long()
+        # -- [B x 2] x,y coordinates
+        coordinates = blstats[:, :2]
+        # TODO ???
+        # coordinates[:, 0].add_(-1)
+
+        # -- [B x F]
+        blstats = blstats.view(T * B, -1).float()
+        # -- [B x K]
+        blstats_emb = self.embed_blstats(blstats)
+
+        assert blstats_emb.shape[0] == T * B
+
+        reps = [blstats_emb]
+
+        # -- [B x H' x W']
+        crop = self.crop(glyphs, coordinates)
+
+        # print("crop", crop)
+        # print("at_xy", glyphs[:, coordinates[:, 1].long(), coordinates[:, 0].long()])
+
+        # -- [B x H' x W' x K]
+        crop_emb = self._select(self.embed, crop)
```

Comment:
Select what? The name should be more informative.
```suggestion
        crop_emb = self._select(self.embed, crop)
```

Reply:
-> FB